File: android-native/app/src/main/java/com/example/kpkn/services/workout/TimerNotificationActionReceiver.kt
package com.example.kpkn.services.workout

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import androidx.core.app.NotificationManagerCompat

class TimerNotificationActionReceiver : BroadcastReceiver() {
    override fun onReceive(context: Context, intent: Intent) {
        if (!ActiveWorkoutHolder.isActive()) return

        when (intent.action) {
            WorkoutRestForegroundService.ACTION_COMPLETE_SET -> {
                ActiveWorkoutHolder.handleAction(TimerAction.CompleteSet)
            }
            WorkoutRestForegroundService.ACTION_SKIP_TIMER -> {
                ActiveWorkoutHolder.handleAction(TimerAction.SkipTimer)
            }
            WorkoutRestForegroundService.ACTION_ADD_TIME -> {
                ActiveWorkoutHolder.handleAction(TimerAction.AddTime)
            }
            WorkoutRestForegroundService.ACTION_SUBTRACT_TIME -> {
                ActiveWorkoutHolder.handleAction(TimerAction.SubtractTime)
            }
        }

        NotificationManagerCompat.from(context).cancel(WorkoutRestForegroundService.NOTIF_ID)
    }
}

File: ios-native/KPKNFit/KPKNFit/Services/Workout/TimerNotificationActionReceiver.swift
import Foundation
import UIKit
import UserNotifications

public enum TimerAction {
    case completeSet
    case skipTimer
    case addTime
    case subtractTime
}

public enum TimerNotificationActionReceiver {

    public static let actionCompleteSet = "TIMER_COMPLETE_SET"
    public static let actionSkipTimer = "TIMER_SKIP"
    public static let actionAddTime = "TIMER_ADD_TIME"
    public static let actionSubtractTime = "TIMER_SUBTRACT_TIME"

    public static let categoryRestTimer = "REST_TIMER"

    public static func registerActions() {
        let complete = UNNotificationAction(identifier: actionCompleteSet, title: "Completar serie", options: [])
        let skip = UNNotificationAction(identifier: actionSkipTimer, title: "Saltar", options: [])
        let add = UNNotificationAction(identifier: actionAddTime, title: "+15s", options: [])
        let subtract = UNNotificationAction(identifier: actionSubtractTime, title: "-15s", options: [])

        let category = UNNotificationCategory(
            identifier: categoryRestTimer,
            actions: [complete, skip, add, subtract],
            intentIdentifiers: [],
            options: []
        )

        UNUserNotificationCenter.current().setNotificationCategories([category])
    }

    public static func handle(response: UNNotificationResponse) {
        guard ActiveWorkoutHolder.shared.isActive() else { return }
 
        switch response.actionIdentifier {
        case actionCompleteSet:
            ActiveWorkoutHolder.shared.handleAction(.completeSet)
        case actionSkipTimer:
            ActiveWorkoutHolder.shared.handleAction(.skipTimer)
        case actionAddTime:
            ActiveWorkoutHolder.shared.handleAction(.addTime)
        case actionSubtractTime:
            ActiveWorkoutHolder.shared.handleAction(.subtractTime)
        default:
            break
        }
    }
}
